crates/polars-plan/src/logical_plan/optimizer/cluster_with_columns.rs

```diff
@@ -0,0 +1,561 @@
+            // input.
+
+            let mut column_map = PlHashMap::default();
+
+            let input_colstat: Vec<ExprColumnStatus> = input_wc
+                .exprs
+                .as_exprs()
+                .iter()
+                .map(|expr| ExprColumnStatus::from_expr_ir(&mut column_map, expr, self.expr_arena))
+                .collect();
+            let current_colstat: Vec<ExprColumnStatus> = current_wc
+                .exprs
+                .as_exprs()
+                .iter()
+                .map(|expr| ExprColumnStatus::from_expr_ir(&mut column_map, expr, self.expr_arena))
+                .collect();
+
+            let child_colstat: Vec<ExprColumnStatusFrozen> = input_colstat
+                .into_iter()
+                .map(|s| s.freeze_with_size(column_map.len()))
+                .collect();
+            let current_colstat: Vec<ExprColumnStatusFrozen> = current_colstat
+                .into_iter()
+                .map(|s| s.freeze_with_size(column_map.len()))
+                .collect();
+
+            // @NOTE: This satisfies the borrow checker
+            let num_names = column_map.len();
+            drop(column_map);
+
+            let mut pushable = MutableBitmap::with_capacity(input_wc.exprs.len());
+
+            let mut input_generate = Bitmap::new_zeroed(num_names);
+            for colstat in &child_colstat {
+                use std::ops::BitOr;
+                input_generate = input_generate.bitor(&colstat.generate);
+            }
+
+            // Check for every expression in the current WITH_COLUMNS node whether
+            for colstat in &current_colstat {
+                use std::ops::BitAnd;
+
+                let has_intersection = input_generate.bitand(&colstat.live).set_bits() > 0;
+                let is_pushable = !has_intersection;
+
+                pushable.push(is_pushable);
+            }
+
+            let pushable = pushable.freeze();
+
+            // There is nothing to push down. Move on.
+            if pushable.set_bits() == 0 {
+                break;
+            }
+
+            // If all columns are pushable, we can merge the input into the current. This should be
+            // a relatively common case.
+            if pushable.set_bits() == pushable.len() {
+                // @NOTE: To keep the schema correct, we reverse the order here. As a
+                // `WITH_COLUMNS` higher up produces later columns. This also allows us not to
+                // have to deal with schemas.
+                input_wc
+                    .exprs
+                    .exprs_mut()
+                    .extend(std::mem::take(current_wc.exprs.exprs_mut()));
+                std::mem::swap(current_wc.exprs.exprs_mut(), input_wc.exprs.exprs_mut());
+
+                *current_wc.input = *input_wc.input;
+                // @TODO: Is this allowed?
+                *current_wc.options = *input_wc.options;
```

Comment:
No, you should take the most conservative options here. You can create a `ProjectionOptions::merge_options`.

Conservative is:
```
duplicate_check: false,
run_parallel: false
```